### PR TITLE
feat(service-worker): introduce the @angular/service-worker package

### PR DESCRIPTION
**Pull request by alxhub**
## Status

This is an in-progress PR, made to facilitate review of the rather large merge to angular/angular.

Remaining tasks before this PR should be merged:

- [x] Unit tests for companion library
- [x] Integration test between companion lib and SW
- [x] Public API generator
- [x] Passing Travis/CircleCI/etc
- [x] ngsw-cli tool in package.json

Remaining tasks that could wait until future PRs, in no particular order:

- [x] Debugging page
- [x] Check for obsolete configuration version (SW internal version or app provided version) on update, and clear prior cache state if discovered
- [x] Check for invalid SW config version
- [x] Kill switch when manifest 404s
- [ ] More unit tests for CLI tool
- [ ] Integration test for CLI tool
- [x] `refreshAhead` behavior for data caching
- [x] Update mode for asset caching, including marking resources as unused
- [ ] Ensure full vs path-only URL matching works correctly
- [x] Support `optimizeFor` behavior

## PR Description

This service worker is a conceptual derivative of the existing @angular/service-worker maintained at github.com/angular/mobile-toolkit, but has been rewritten to support use across a much wider variety of applications.

Entrypoints include:

@angular/service-worker: a library for use within Angular client apps to communicate with the service worker.
@angular/service-worker/gen: a library for generating ngsw.json files from glob-based SW config files.
@angular/service-worker/ngsw-worker.js: the bundled service worker script itself.
@angular/service-worker/ngsw-cli.js: a CLI tool for generating ngsw.json files from glob-based SW config files.

### CODE REVIEW

**Comment by Ploppy3** [comment]
I hope you fixed this: https://github.com/angular/mobile-toolkit/issues/150

File: tools/public_api_guard/service-worker/gen.d.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    constructor(fs: Filesystem, baseHref: string);
+    process(config: NgswConfig): Promise<Object>;
+    processAssetGroups(config: NgswConfig, hashTable: {
+        [file: string]: string | undefined;
+    }): Promise<Object[]>;
+    processDataGroups(config: NgswConfig): Object[];
+}
+
+/** @experimental */
+export interface NgswConfig {
```
Comment (IgorMinar) [inline tools/public_api_guard/service-worker/gen.d.ts]: sw

File: tools/public_api_guard/service-worker/config.d.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    readonly fs: Filesystem;
+    constructor(fs: Filesystem, baseHref: string);
+    process(config: NgswConfig): Promise<Object>;
+}
+
+/** @experimental */
+export declare type Glob = string;
+
+/** @experimental */
+export interface NgswConfig {
```
Comment (IgorMinar) [inline tools/public_api_guard/service-worker/config.d.ts]: sw

File: packages/service-worker/test/integration_spec.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import 'rxjs/add/operator/take';
```
Comment (IgorMinar) [inline packages/service-worker/test/integration_spec.ts]: we should not be using these because they could affect other tests

**Review by IgorMinar — APPROVED** [review]

**Comment by alexjlockwood** [comment]
Hi @alxhub, is there any documentation anywhere that explains how to upgrade the older beta service-worker package to the newer 5.0.0+ version?

**Comment by Ploppy3** [comment]
@alxhub We need more info on this.

**Pull request merged**